### PR TITLE
Staleness for rules

### PR DESCRIPTION
**Pull request by brian-brazil**
This should work with no changes once we have rule groups. If a rule is deleted we don't add stale markers, but that's okay as it no longer exists (or is in another group now) so if it's stale for a bit that's fine.

Note there's a semantic change in the last commit.

@fabxc

### CODE REVIEW

File: rules/manager.go [hunk 1]
```diff
@@ -214,25 +220,41 @@ func (g *Group) offset() time.Duration {
...
 func (g *Group) copyState(from *Group) {
-	for _, fromRule := range from.rules {
-		far, ok := fromRule.(*AlertingRule)
+	ruleMap := make(map[string][]int, len(from.rules))
+
+	for fi, fromRule := range from.rules {
+		l, ok := ruleMap[fromRule.Name()]
 		if !ok {
+			l = []int{}
+		}
```
Comment (fabxc) [inline rules/manager.go]: This can be left out as `append(nil, fi)` is valid.

File: rules/manager_test.go [hunk 2]
```diff
@@ -156,3 +156,143 @@ func annotateWithTime(lines []string, ts time.Time) []string {
...
+	app, _ = storage.Appender()
+	app.Add(labels.FromStrings(model.MetricNameLabel, "a"), 1000, 2)
+	if err = app.Commit(); err != nil {
+		t.Fatal(err)
+	}
+	app, _ = storage.Appender()
+	app.Add(labels.FromStrings(model.MetricNameLabel, "a"), 2000, math.Float64frombits(value.StaleNaN))
+	if err = app.Commit(); err != nil {
+		t.Fatal(err)
+	}
```
Comment (fabxc) [inline rules/manager_test.go]: No need to open 3 appenders. This can all go into a single transaction.

File: rules/manager_test.go [hunk 3]
```diff
@@ -156,3 +156,143 @@ func annotateWithTime(lines []string, ts time.Time) []string {
...
+	group.Eval(time.Unix(2, 0))
+
+	querier, err := storage.Querier(0, 2000)
+	defer querier.Close()
+	if err != nil {
+		t.Fatal(err)
+	}
+	matcher, _ := labels.NewMatcher(labels.MatchEqual, model.MetricNameLabel, "a_plus_one")
+	seriesSet := querier.Select(matcher)
+	samples, err := readSeriesSet(seriesSet)
```
Comment (fabxc) [inline rules/manager_test.go]: Can go into a single line: `samples, err := readSeriesSet(querier.Select(matcher))`

File: rules/manager_test.go [hunk 4]
```diff
@@ -156,3 +156,143 @@ func annotateWithTime(lines []string, ts time.Time) []string {
...
+			points = append(points, promql.Point{T: t, V: v})
+		}
+
+		name := series.Labels().String()
+		result[name] = points
+		if err := ss.Err(); err != nil {
+			return nil, err
+		}
+	}
+	return result, nil
```
Comment (fabxc) [inline rules/manager_test.go]: `return result, ss.Err()`. The check shouldn't go into the loop as `ss.Next()` will exit it already if there is an error.

**Comment by fabxc** [comment]
Looks okay aside from code nits.

**Comment by brian-brazil** [comment]
Updated, ptal.

**Comment by fabxc** [comment]
👍

**Pull request merged**